py/dml/codegen.py

```diff
@@ -2034,74 +2055,75 @@ def stmt_for(stmt, location, scope):
-            or not isinstance(stmts[0], (ctree.Case, ctree.Default))):
-            raise ESWITCH(
-                body_ast.site, "statement before first case label")
-        defaults = [i for (i, sub) in enumerate(stmts)
-                    if isinstance(sub, ctree.Default)]
-        if len(defaults) > 1:
-            raise ESWITCH(stmts[defaults[1]].site, "duplicate default label")
-        if defaults:
-            for sub in stmts[defaults[0]:]:
-                if isinstance(sub, ctree.Case):
-                    raise ESWITCH(sub.site,
-                                  "case label after default label")
-        body = ctree.Compound(body_ast.site, stmts)
-    else:
-        body = codegen_statement(body_ast, location, scope)
-    try:
+    with CLoopContext():
+        if stmt.site.dml_version() != (1, 2):
+            assert body_ast.kind == 'compound'
+            [stmt_asts] = body_ast.args
+            stmts = codegen_statements(stmt_asts, location, scope)
+            if (not stmts
+                or not isinstance(stmts[0], (ctree.Case, ctree.Default))):
+                raise ESWITCH(
+                    body_ast.site, "statement before first case label")
+            defaults = [i for (i, sub) in enumerate(stmts)
+                        if isinstance(sub, ctree.Default)]
+            if len(defaults) > 1:
+                raise ESWITCH(stmts[defaults[1]].site,
+                              "duplicate default label")
+            if defaults:
+                for sub in stmts[defaults[0]:]:
+                    if isinstance(sub, ctree.Case):
+                        raise ESWITCH(sub.site,
+                                      "case label after default label")
+            body = ctree.Compound(body_ast.site, stmts)
+        else:
+            body = codegen_statement(body_ast, location, scope)
+
         res = mkSwitch(stmt.site, expr, body)
-    finally:
-        loop_stack.pop()
     return [res]
 
 @statement_dispatcher
 def stmt_continue(stmt, location, scope):
-    if not loop_stack or loop_stack[-1] == 'method':
+    if (LoopContext.current is None
+        or isinstance(LoopContext.current, NoLoopContext)):
         raise ECONT(stmt.site)
-    elif loop_stack[-1] == 'c':
+    elif isinstance(LoopContext.current, CLoopContext):
         return [mkContinue(stmt.site)]
     else:
         raise ECONTU(stmt.site)
 
 @statement_dispatcher
 def stmt_break(stmt, location, scope):
-    if not loop_stack or loop_stack[-1] == 'method':
+    if (LoopContext.current is None
+        or isinstance(LoopContext.current, NoLoopContext)):
         raise EBREAK(stmt.site)
-    elif loop_stack[-1] == 'c':
+    elif isinstance(LoopContext.current, CLoopContext):
         return [mkBreak(stmt.site)]
+    elif isinstance(LoopContext.current, UnrolledLoopContext):
+        LoopContext.current.used = True
+        return [mkUnrolledBreak(stmt.site,
+                                f'_unrolledbreak{LoopContext.current.id}')]
```

Comment:
We could consider making this a method on LoopContext, `return LoopContext.current.break_()`

Reply:
I very slightly prefer keeping the code in `stmt_break`, but I don't have strong feelings about it.

Reply:
It would eliminate the `else` clause.

Also, would make sense to store the full label `_unrolledbreak{id}` instead of the ID in LoopContext?

Reply:
> It would eliminate the else clause.

I suppose that's a good enough argument to go with that route.

> Also, would make sense to store the full label _unrolledbreak{id} instead of the ID in UnrolledLoopContext?

I'd much sooner prefer that `UnrolledBreak.toc()` adds `_unrolledbreak` and `mkUnrolledBreak` just gets passed the ID.

Reply:
The identifier is written in two places (goto and label); it would be good if we didn't duplicate this string in two classes. (ok to repeat it twice within one class)

Reply:
Ah, yes, that makes sense.